### PR TITLE
Store-gateway: allow mark/delete no-compaction from admin UI

### PR DESCRIPTION
**Pull request by ying-jeanne**
#### What this PR does
Compactor is sometimes unable to compact blocks because they are too big, until we increases compactor memory requested manually, we often mark for no compact the specific block. (there are also other cases we mark for no compact blocks when data corrupted) This PR simplifies this operation.

Allow admin to mark/delete no-compact from admin UI, followup of: https://github.com/grafana/mimir/pull/6917 and https://github.com/grafana/mimir/pull/6848. 
Note metric `cortex_compactor_blocks_marked_for_no_compaction_total` was not passed to block.go intentionally. This metric is excluded because normally we rely on it for monitoring the number of compaction that have failed, and in this context, the marking or deletion of "no-compact" status is a intentional and manual action.

The local test
output

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [x] [`about-versioning.md`](/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/storegateway/blocks.gohtml [hunk 1]
```diff
@@ -90,5 +103,82 @@
...
+                continue;
+            }
+            // Add the uid to the array
+            blockUIDs.push(cells[1].textContent);
+        }
+
+        // Check if the array is empty
+        if (blockUIDs.length === 0) {
+            // Display an alert popup if the array is empty
+            alert("No block seleted for action!");
```
Comment (dimitarvdimitrov) [inline pkg/storegateway/blocks.gohtml:160]:
```suggestion
            alert("Select at least one block and then perform the action");
```

this sounds a bit more positive

File: pkg/storegateway/blocks.gohtml [hunk 2]
```diff
@@ -90,5 +103,82 @@
...
+        }
+
+        // Check if the input field already exists
+        var blockUIDsField = form.querySelector('input[name="block_uids"]');
```
Comment (dimitarvdimitrov) [inline pkg/storegateway/blocks.gohtml:164]: technically we call them ULIDs or IDs, but not UIDs. Let's stick to ULIDs, wdyt?

File: pkg/storegateway/blocks.gohtml [hunk 3]
```diff
@@ -90,5 +103,82 @@
...
+        toggleVisibility(isVisible);
+    });
+
+    // Set the initial visibility based on the selected option
+    var isVisible = dropdown.value !== "none";
+    toggleVisibility(isVisible);
+
+    // Function to reload the page with the action
+    function reloadWithAction() {
+        var form = document.getElementById('configForm');
```
Comment (dimitarvdimitrov) [inline pkg/storegateway/blocks.gohtml:135]: should we move this closer to where its used?

File: pkg/storage/tsdb/block/block.go [hunk 4]
```diff
@@ -338,9 +338,8 @@ func MarkForNoCompact(ctx context.Context, logger log.Logger, bkt objstore.Bucke
 	}
```
Comment (dimitarvdimitrov) [inline pkg/storage/tsdb/block/block.go:339]: this function now logs `requested to mark for no compaction, but file already exists; this should not happen; investigate` when you try to mark a block twice. I think this is confusing. 

One way to rework is to return a special sentinel error from the block package and handle that differently in the store-gateway and compactor packages.

File: pkg/storegateway/gateway_blocks_http.go [hunk 5]
```diff
@@ -74,6 +97,73 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
+		}
+		for _, uid := range uids {
+			ulid, err := ulid.Parse(uid)
+			if err != nil {
+				util.WriteTextResponse(w, fmt.Sprintf("Can't parse ULID %s: %s", uid, err))
+				return
+			}
+			bkt := block.BucketWithGlobalMarkers(bucket.NewUserBucketClient(tenantID, s.stores.bucket, nil))
+			if action == ActionTypeNoCompact {
+				err = block.MarkForNoCompact(
```
Comment (dimitarvdimitrov) [inline pkg/storegateway/gateway_blocks_http.go:139]: let's add some logging trace that this marker was uploaded by some user. Perhaps adding a special logging line in this function.

in addition to the tenant ID if we can get their IP that would be great, but I'm not sure how easy it will be. Don't want to block the PR because of the IP

File: pkg/storegateway/gateway_blocks_http.go [hunk 6]
```diff
@@ -74,6 +97,73 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
+
+		decodedBytes, err := base64.StdEncoding.DecodeString(decodedString)
+		if err != nil {
+			util.WriteTextResponse(w, fmt.Sprintf("Can't decode base64 of selected blocks' uid: %s", err))
+			return
+		}
+		if err := json.Unmarshal(decodedBytes, &uids); err != nil {
+			util.WriteTextResponse(w, fmt.Sprintf("Can't parse block_uids: %s", err))
+			return
+		}
```
[the remark this answers is not in the document]
Comment (ying-jeanne) [reply pkg/storegateway/gateway_blocks_http.go:130]: what do you mean the same code elsewhere? this is a json struct i encoded on frontend, not used anywhere else.

File: pkg/storage/tsdb/block/block.go (hunk 4 above)
Comment (ying-jeanne) [reply pkg/storage/tsdb/block/block.go:339]: I might not making it an error, this is a warning log, the remark is not crucial failure - the current compactor should continue the compaction, not trigger any alert because of this. Since it would just end up on a no-op not worthy an error. The best I can do is to pass a errorhandler function, for compactor log warning and return nil error, but for store-gateway return error on frontend. But does it worth it for the code change?

File: pkg/storegateway/gateway_blocks_http.go (hunk 5 above)
Comment (ying-jeanne) [reply pkg/storegateway/gateway_blocks_http.go:139]: added.

File: pkg/storage/tsdb/block/block.go (hunk 4 above)
Comment (dimitarvdimitrov) [reply pkg/storage/tsdb/block/block.go:339]: can we just return an error which the store-gateway ignores and the compactor always logs. Something like


```go 
var MarkExists = fmt.Errorf("mark exists")

func MarkForNoCompact() error {
	...
	if noCompactMarkExists {
		return MarkExists
	}
	...
}
```

compactor:
```go
if errors.Is(err, block.MarkExists) {
	log("requested to mark for no compaction, but file already exists; this should not happen; investigate")
}
```

and 

store-gateway:
```go
if !errors.Is(err, block.MarkExists) {
	log("couldn't mark block for no compaction")
}
```

Comment (ying-jeanne) [reply pkg/storage/tsdb/block/block.go:348]: I personally don't like to return error and abandon the mark for following blocks just because of a no-op, but I don't want to block the PR just for this. updated as requested.

**Comment by ying-jeanne** [comment]
the block_ulids is trasmit as json object, tested with frontend and backend, since it is URL-safe, no encoding needed.

File: pkg/storage/tsdb/block/block.go (hunk 4 above)
Comment (dimitarvdimitrov) [reply pkg/storage/tsdb/block/block.go:348]: > return error and abandon the mark for following blocks just because of a no-op

that's a good point, I don't think we should be changing the behaviour. How about we use a multiError to collect these errors? [for example here](https://github.com/grafana/mimir/blob/15ddfa77995aec1285ccb5f217861dbc094ebfe4/pkg/ruler/ruler.go#L1286)

Comment (ying-jeanne) [reply pkg/storage/tsdb/block/block.go:348]: I like this multiError, added the change and + a small improvement on frontend.

File: pkg/storegateway/gateway_blocks_http.go [hunk 7]
```diff
@@ -156,9 +207,43 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
 		ShowSources: showSources,
 		ShowParents: showParents,
+		ActionType:  action,
 	}, blocksPageTemplate, req)
 }
 
+func (s *StoreGateway) performActionsOnBlocks(tenantID string, req *http.Request, action ActionType, blockUlids []string) error {
+	// When blockUlids is set, and dropdown action is "no-compact" or "delete-no-compact",
+	// we will perform the action on the selected blocks
+	if (action == ActionTypeNoCompact || action == ActionTypeDeleteNoCompact) && len(blockUlids) > 0 {
```
Comment (dimitarvdimitrov) [inline pkg/storegateway/gateway_blocks_http.go:217]: we do the validation for the action and length of block IDs here, but also in `BlocksHandler`. Let's do it in one place only.

**Review by dimitarvdimitrov — COMMENTED** [review]
mostly LGTM. I left nitpicks and minor comments. The gist of it is i think the error handling can be simplified and error logs can be slightly improved

File: pkg/storegateway/gateway_blocks_http.go [hunk 8]
```diff
@@ -156,9 +207,43 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
+		errs := multierror.MultiError{}
+		for _, uid := range blockUlids {
+			ulid, err := ulid.Parse(uid)
+			if err != nil {
+				return fmt.Errorf("can't parse ULID %s: %w", uid, err)
+			}
+			bkt := block.BucketWithGlobalMarkers(bucket.NewUserBucketClient(tenantID, s.stores.bucket, nil))
+			switch action {
+			case ActionTypeNoCompact:
+				errs.Add(block.MarkForNoCompact(req.Context(), s.logger, bkt, ulid, block.ManualNoCompactReason, "Manual Operations from Admin UI: Mark for no compaction", nil))
```
[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/storegateway/gateway_blocks_http.go:227]: on a second thought i think it's ok as it is

File: pkg/storegateway/gateway_blocks_http.go (hunk 7 above)
Comment (dimitarvdimitrov) [reply pkg/storegateway/gateway_blocks_http.go:217]: note from @ying-jeanne the check for the empty string in BlocksHanlder and the check here are different - a non-empty string may still produce empty slice

File: pkg/storegateway/gateway_blocks_http.go [hunk 9]
```diff
@@ -74,6 +98,33 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
 	showDeleted := req.Form.Get("show_deleted") == "on"
 	showSources := req.Form.Get("show_sources") == "on"
 	showParents := req.Form.Get("show_parents") == "on"
+	actionType := req.Form.Get("action_type")
```
[the remark this answers is not in the document]
Comment (ying-jeanne) [reply pkg/storegateway/gateway_blocks_http.go:101]: because when it is empty, it would be default to none. just wanted to keep it as enum, do you mean we keep action as string and just hard coded == instead?

File: pkg/compactor/bucket_compactor.go [hunk 10]
```diff
@@ -857,7 +860,10 @@ func (c *BucketCompactor) Compact(ctx context.Context, maxCompactionTime time.Du
 							"UnhealthyBlock: marking unhealthy block as no compact to unblock compaction",
 							c.metrics.blocksMarkedForNoCompact.WithLabelValues(block.CriticalNoCompactReason),
 						)
-						if err == nil {
+						if err != nil && errors.As(err, &block.TSDBBlockRecoverableError{}) {
```
[the remark this answers is not in the document]
Comment (ying-jeanne) [reply pkg/compactor/bucket_compactor.go:872]: it is needed because of this `As will panic if target is not a non-nil pointer to either a type that implements error, or to any interface type. As returns false if err is nil.` the unittest failed, that is why i added the error check

**Comment by duricanikolic** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

File: pkg/storegateway/gateway_blocks_http.go [hunk 11]
```diff
@@ -156,9 +207,43 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
+			default:
+				return nil
+			}
+		}
+		ip := req.Header.Get("X-Forwarded-For")
+		// If X-Forwarded-For is empty, fall back to RemoteAddr
+		if ip == "" {
+			ip = strings.Split(req.RemoteAddr, ":")[0]
+		}
+		level.Info(s.logger).Log("msg", "Performed action on blocks", "action", action, "blocks", blockUlids, "ip", ip)
```
[the remark this answers is not in the document]
Comment (ying-jeanne) [reply pkg/storegateway/gateway_blocks_http.go]: added.

File: pkg/storage/tsdb/block/block.go [hunk 12]
```diff
@@ -42,6 +43,14 @@ const (
 	DebugMetas = "debug/metas"
 )
 
+type ErrMarkerExists struct {
+	message string
+}
+
+func (e *ErrMarkerExists) Error() string {
```
Comment (colega) [inline pkg/storage/tsdb/block/block.go:50]:
```suggestion
func (e ErrMarkerExists) Error() string {
```
Should not be a pointer, as it makes it's usage cumbersome.
Note that if it's a pointer, then the above call `errors.As(err, &block.ErrMarkerExists{})` will never be true.

File: pkg/storegateway/blocks.gohtml [hunk 13]
```diff
@@ -90,5 +103,92 @@
     {{ end }}
     </tbody>
 </table>
+<script>
```
Comment (colega) [inline pkg/storegateway/blocks.gohtml:106]: Maybe we could use something like htmlx.org?

File: pkg/storegateway/gateway_blocks_http.go [hunk 14]
```diff
@@ -156,9 +207,45 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
 		ShowDeleted: showDeleted,
 		ShowSources: showSources,
 		ShowParents: showParents,
+		ActionType:  action,
 	}, blocksPageTemplate, req)
 }
 
+func (s *StoreGateway) performActionsOnBlocks(tenantID string, req *http.Request, action ActionType, blockUlids []string) error {
```
Comment (colega) [inline pkg/storegateway/gateway_blocks_http.go:214]: I see how convenient is to implement it this way, but I think this is the first operation where store-gateway isn't a reader of block storage, but also a writer of it.

Are we good with that?

(I think this better fits compactor, who's the one writing and dealing with compaction, but I also understand the inconvenience of having it in a different component).

File: pkg/storegateway/gateway_blocks_http.go [hunk 15]
```diff
@@ -74,6 +98,33 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
+
+	if !isValidActionType(ActionType(actionType)) {
+		util.WriteTextResponse(w, fmt.Sprintf("Invalid Action Type: %s\n", actionType))
+		return
+	}
+	action := ActionType(actionType)
+
+	blockUlidsString := req.Form.Get("block_ulids")
+	if action != ActionTypeNone && blockUlidsString != "" {
+		var uids []string
```
Comment (colega) [inline pkg/storegateway/gateway_blocks_http.go:114]:
```suggestion
		var ulids []string
```

File: pkg/storegateway/gateway_blocks_http.go [hunk 16]
```diff
@@ -74,6 +98,33 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
+		if err != nil {
+			util.WriteTextResponse(w, fmt.Sprintf("Can't decode base64 of selected blocks' uid: %s", err))
+			return
+		}
+		err = s.performActionsOnBlocks(tenantID, req, action, uids)
+		if err != nil {
+			util.WriteTextResponse(w, err.Error())
+			return
+		}
+	}
```
Comment (colega) [inline pkg/storegateway/gateway_blocks_http.go:126]: If we didn't fail, it would be nice to provide some kind of feedback here.

File: pkg/storegateway/gateway_blocks_http.go [hunk 17]
```diff
@@ -74,6 +98,33 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
+	actionType := req.Form.Get("action_type")
+	if actionType == "" {
+		actionType = string(ActionTypeNone) // Set the default value to "none"
+	}
+
+	if !isValidActionType(ActionType(actionType)) {
+		util.WriteTextResponse(w, fmt.Sprintf("Invalid Action Type: %s\n", actionType))
+		return
+	}
+	action := ActionType(actionType)
```
Comment (colega) [inline pkg/storegateway/gateway_blocks_http.go:110]: We should make sure here that this is a POST action.

File: pkg/compactor/bucket_compactor.go [hunk 18]
```diff
@@ -913,7 +916,10 @@ func (c *BucketCompactor) Compact(ctx context.Context, maxCompactionTime time.Du
 							"UnhealthyBlock: marking unhealthy block as no compact to unblock compaction",
 							c.metrics.blocksMarkedForNoCompact.WithLabelValues(block.CriticalNoCompactReason),
 						)
-						if err == nil {
+						if errors.As(err, &block.ErrMarkerExists{}) {
+							level.Warn(c.logger).Log("msg", "mark block is skipped for no critical reason, this should not happen", "err", err)
+						}
+						if err == nil || errors.As(err, &block.ErrMarkerExists{}) {
```
[the remark this answers is not in the document]
Comment (colega) [reply pkg/compactor/bucket_compactor.go:922]: (Suggestion looks weird, but I'd move the `errors.As` condition into `err == nil || errors.As` condition block)

[the remark this answers is not in the document]
Comment (colega) [reply pkg/compactor/bucket_compactor.go:922]: Same applies above.

[the remark this answers is not in the document]
Comment (ying-jeanne) [reply pkg/compactor/bucket_compactor.go]: i think this is valid, since we will have less request pass through this condition, may save us some time.

File: pkg/storegateway/blocks.gohtml (hunk 13 above)
Comment (ying-jeanne) [reply pkg/storegateway/blocks.gohtml:114]: for the simple submit i replace with htmx function, for the post, i need to parse the table and get the list of ulid to send, for that I kept the scripts.

File: pkg/storegateway/gateway_blocks_http.go (hunk 16 above)
Comment (ying-jeanne) [reply pkg/storegateway/gateway_blocks_http.go]: added info text, to clarify what are changed, how many blocks modified

File: pkg/storage/tsdb/block/block.go [hunk 19]
```diff
@@ -335,22 +344,21 @@ func GetMetaAttributes(ctx context.Context, meta *Meta, bucketReader objstore.Bu
...
 func MarkForNoCompact(ctx context.Context, logger log.Logger, bkt objstore.Bucket, id ulid.ULID, reason NoCompactReason, details string, markedForNoCompact prometheus.Counter) error {
 	m := path.Join(id.String(), NoCompactMarkFilename)
 	noCompactMarkExists, err := bkt.Exists(ctx, m)
 	if err != nil {
 		return errors.Wrapf(err, "check exists %s in bucket", m)
 	}
 	if noCompactMarkExists {
-		level.Warn(logger).Log("msg", "requested to mark for no compaction, but file already exists; this should not happen; investigate", "err", errors.Errorf("file %s already exists in bucket", m))
-		return nil
+		return &ErrMarkerExists{
```
Comment (colega) [inline pkg/storage/tsdb/block/block.go:355]:
```suggestion
		return ErrMarkerExists{
```

File: pkg/storegateway/gateway_blocks_http.go (hunk 14 above)
Comment (ying-jeanne) [reply pkg/storegateway/gateway_blocks_http.go]: as discussed in the weekly, we can leave this in store-gateway for now, if in the future we want to remove the write permission of store-gateway, I think this is acceptable since we are doing revertable changes, upload/delete no-compact marker is not a critical action.

File: pkg/storegateway/gateway_blocks_http.go [hunk 20]
```diff
@@ -156,9 +207,43 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
+	// When blockUlids is set, and dropdown action is "no-compact" or "delete-no-compact",
+	// we will perform the action on the selected blocks
+	if (action == ActionTypeNoCompact || action == ActionTypeDeleteNoCompact) && len(blockUlids) > 0 {
+		errs := multierror.MultiError{}
+		for _, uid := range blockUlids {
+			ulid, err := ulid.Parse(uid)
+			if err != nil {
+				return fmt.Errorf("can't parse ULID %s: %w", uid, err)
+			}
+			bkt := block.BucketWithGlobalMarkers(bucket.NewUserBucketClient(tenantID, s.stores.bucket, nil))
```
[the remark this answers is not in the document]
Comment (ying-jeanne) [reply pkg/storegateway/gateway_blocks_http.go]: for adminUI, it do only one tenant, won't make any change.

[the remark this answers is not in the document]
Comment (ying-jeanne) [reply pkg/storegateway/gateway_blocks_http.go]: moved bucket client creation out of loop

File: pkg/storegateway/blocks.gohtml [hunk 21]
```diff
@@ -4,25 +4,45 @@
...
 <body>
 <h1>Store-gateway: bucket tenant blocks</h1>
 <p>Current time: {{ .Now }}</p>
 <p>Showing blocks for tenant: <strong>{{ .Tenant }}</strong></p>
 <p>
-    <form>
-        <input type="checkbox" id="show-deleted" name="show_deleted" {{ if .ShowDeleted }} checked {{ end }}>&nbsp;<label for="show-deleted">Show Deleted</label> &nbsp;&nbsp;
-        <input type="checkbox" id="show-sources" name="show_sources" {{ if .ShowSources }} checked {{ end }}>&nbsp;<label for="show-sources">Show Sources</label> &nbsp;&nbsp;
-        <input type="checkbox" id="show-parents" name="show_parents" {{ if .ShowParents }} checked {{ end }}>&nbsp;<label for="show-parents">Show Parents</label> &nbsp;&nbsp;
+    <form id="configForm">
```
Comment (colega) [inline pkg/storegateway/blocks.gohtml:14]: Nit: rest of the elements in the html have dashes in their names, can this be `config-form`?

File: pkg/storegateway/blocks.gohtml [hunk 22]
```diff
@@ -4,25 +4,45 @@
...
+                hx-trigger="click"
+                hx-target="body"
+                hx-include="#configForm"
+                hx-push-url="true">
+            <span style="padding: 0.5em 1em; font-size: 100%;">Reload</span>
         </button>
     </form>
 </p>
-<table border="1" cellpadding="5" style="border-collapse: collapse">
+<table border="1" cellpadding="5" style="border-collapse: collapse" id="blocksInfo">
```
Comment (colega) [inline pkg/storegateway/blocks.gohtml:42]: Same here, `blocks-info`?

File: pkg/storegateway/gateway_blocks_http.go [hunk 23]
```diff
@@ -156,9 +197,105 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
+	tenantID := vars["tenant"]
+	if tenantID == "" {
+		util.WriteTextResponse(w, "Tenant ID can't be empty")
+		return
+	}
+
+	if err := req.ParseForm(); err != nil {
+		util.WriteTextResponse(w, fmt.Sprintf("Can't parse form: %s", err))
+		return
+	}
```
Comment (colega) [inline pkg/storegateway/gateway_blocks_http.go:216]: What's the status code of these responses? Looks like they should be at least a 400, but method call says  just "write text response".

File: pkg/storegateway/gateway_blocks_http.go [hunk 24]
```diff
@@ -156,9 +197,105 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
+		actionType = string(ActionTypeNone) // Set the default value to "none"
+	}
+
+	if !isValidActionType(ActionType(actionType)) {
+		util.WriteTextResponse(w, fmt.Sprintf("Invalid Action Type: %s\n", actionType))
+		return
+	}
+
+	// Perform the action on selected blocks
+	action := ActionType(actionType)
```
Comment (colega) [inline pkg/storegateway/gateway_blocks_http.go:228]: Opinionated, but I find it weird that we convert a typed const to a string, and then we validate that it's the correct value.

How about we only validate when there's input?

```suggestion
	action := ActionTypeNone
	if actionType := req.Form.Get("action_type"); actionType != "" {
		if !isValidActionType(ActionType(actionType)) {
			util.WriteTextResponse(w, fmt.Sprintf("Invalid Action Type: %s\n", actionType))
			return
		}	
		action = ActionType(actionType)
	}
```

Secondary nit is: `isValidActionType` should accept a `string`, not an `ActionType`, this way we can put boundaries and assume that anything of type `ActionType` is valid.

File: pkg/storegateway/gateway_blocks_http.go [hunk 25]
```diff
@@ -156,9 +197,105 @@ func (s *StoreGateway) BlocksHandler(w http.ResponseWriter, req *http.Request) {
...
+
+	blockUlidsString := req.Form.Get("block_ulids")
+	var infoText string
+	if action != ActionTypeNone && blockUlidsString != "" {
```
Comment (colega) [inline pkg/storegateway/gateway_blocks_http.go:232]: Is it valid to call the action endpoint without an action? What's the point? If it is, I'd at least add a comment here.

File: pkg/storegateway/blocks.gohtml [hunk 26]
```diff
@@ -90,5 +111,82 @@
...
+        var isVisible = dropdown.value !== "none";
+        toggleVisibility(isVisible);
+    });
+
+    // Set the initial visibility based on the selected option
+    var isVisible = dropdown.value !== "none";
+    toggleVisibility(isVisible);
+
+    // Function to reload the page with the action
+    function reloadWithAction() {
```
Comment (colega) [inline pkg/storegateway/blocks.gohtml:141]: This function submits the form and performs the action, it doesn't just reload.

It should be called `submitForm` or maybe `submitFormAndReload` if you want to underline the fact that it will reload.

File: pkg/storegateway/blocks.gohtml [hunk 27]
```diff
@@ -90,5 +111,82 @@
...
+        // Check if the array is empty
+        if (blockULIDs.length === 0) {
+            // Display an alert popup if the array is empty
+            alert("Select at least one block and then perform the action");
+        }
+        var form = document.getElementById('configForm');    
+        form.querySelector('input[name="block_ulids"]').value = JSON.stringify(blockULIDs);
+        form.method = 'POST';
+        form.submit();
+    }
```
Comment (colega) [inline pkg/storegateway/blocks.gohtml:171]: Can you explain why all of this is needed? Why can't we handle the form _in the normal html way_?

Comment (ying-jeanne) [reply pkg/storegateway/blocks.gohtml:171]: I might have overlooked a simpler method. Since we can select checkboxes in the table, the selected ULIDs should be sent with the request for deletion or updates. However, because the table isn't part of the form, I need to create a variable to parse the table and retrieve the selected ULIDs.

Comment (colega) [reply pkg/storegateway/blocks.gohtml]: What is preventing us from making the table part of the form?

File: pkg/storegateway/gateway_blocks_http.go (hunk 25 above)
Comment (colega) [reply pkg/storegateway/gateway_blocks_http.go]: I would fail with a 400 if someone calls the write endpoint without wanting to write something.

File: pkg/storegateway/blocks.gohtml (hunk 27 above)
Comment (ying-jeanne) [reply pkg/storegateway/blocks.gohtml]: we may have a big request body if don't preprocess the table data when we select show parents/ show delete.

Comment (colega) [reply pkg/storegateway/blocks.gohtml]: Only the checked boxes are sent in the form (I obviously had to check this).

Even if all of them were sent, with 26 bytes per ulid, plus the checkbox name, we'd have to send hundreds of thousands of block IDs back to be an issue (assuming 10MB is a reasonable POST). I think your browser will crash before the request size becomes an issue, and you'll have to fallback to the CLI anyway.

Comment (ying-jeanne) [reply pkg/storegateway/blocks.gohtml]: If we only want checkbox + ulid in the request, we will have to customerise the submit anyway. I was just not convinced it is a nice API design, but if you think it is a cleaner solution to move maximum logics from frontend to backend with all information in the request. I can change to that.

Comment (colega) [reply pkg/storegateway/blocks.gohtml]: > If we only want checkbox + ulid in the request, we will have to customerise the submit anyway. 

Sorry I did not understand you here, can you rephrase please?

> If you think it is a cleaner solution to move maximum logics from frontend to backend with all information in the request. I can change to that.

I'd prefer to have the least possible logic in the frontend as we're not testing it.

**Comment by colega** [comment]
Also one comment: regarding "actions to perform", I would rather have separate buttons for each action instead of a dropdown. You really want to click on "DELETE", rather than clicking on "Execute" and realizing that the action selected wasn't the one you wanted.
